### PR TITLE
List/room granularity for the receipts extension 

### PR DESCRIPTION
**Pull request by DMRobertson**
Part of #41. Based on #53.

### CODE REVIEW

File: testutils/m/match.go [hunk 1]
```diff
@@ -539,6 +548,59 @@ func MatchAccountData(globals []json.RawMessage, rooms map[string][]json.RawMess
...
+				return fmt.Errorf("MatchNoRoomAccountData: got account data for %s, but expected it to be missing", roomID)
+			}
+		}
+		return nil
+	}
+}
+
+// LogResponse builds a matcher that always succeeds. As a side-effect, it pretty-prints
+// the given sync response to the test log. This is useful when debugging a test.
+func LogResponse(t *testing.T) RespMatcher {
```
Comment (kegsay) [inline testutils/m/match.go:596]: Fancy, I like it!

File: tests-e2e/receipts_test.go [hunk 2]
```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
...
+	alice := registerNewUser(t)
+	bob := registerNewUser(t)
+
+	var syncResp *sync3.Response
+
+	t.Log("Alice creates four rooms.")
+	room1 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 1"})
+	room2 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 2"})
+	room3 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 1"})
+	room4 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 2"})
```
Comment (kegsay) [inline tests-e2e/receipts_test.go:236]:
```suggestion
	room4 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 4"})
```

File: tests-e2e/receipts_test.go [hunk 3]
```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
...
+func TestReceiptsRespectsExtensionScope(t *testing.T) {
+	alice := registerNewUser(t)
+	bob := registerNewUser(t)
+
+	var syncResp *sync3.Response
+
+	t.Log("Alice creates four rooms.")
+	room1 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 1"})
+	room2 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 2"})
+	room3 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 1"})
```
Comment (kegsay) [inline tests-e2e/receipts_test.go:235]:
```suggestion
	room3 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 3"})
```

File: tests-e2e/receipts_test.go [hunk 4]
```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
...
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Hello, room!",
+		},
+	}
+	message1 := alice.SendEventSynced(t, room1, messageEvent)
+	message2 := alice.SendEventSynced(t, room1, messageEvent)
+	message3 := alice.SendEventSynced(t, room1, messageEvent)
+	message4 := alice.SendEventSynced(t, room1, messageEvent)
```
Comment (kegsay) [inline tests-e2e/receipts_test.go:256]: Presumably meant to be rooms 1-4 not room 1 all the time.

File: tests-e2e/receipts_test.go [hunk 5]
```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
...
+
+	t.Log("Bob posts a public read receipt for the messages in rooms 1 and 2.")
+	bob.SendReceipt(t, room1, message1, "m.read")
+	bob.SendReceipt(t, room2, message2, "m.read")
+	time.Sleep(300 * time.Millisecond) // TODO: find a better way to wait until the proxy has processed this.
```
Comment (kegsay) [inline tests-e2e/receipts_test.go:261]: You've not done a sliding sync request yet, so this sleep means nothing.

File: tests-e2e/receipts_test.go [hunk 6]
```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
...
+	}
+
+	go bobIncrementalSync()
+	t.Log("Bob posts receipts for rooms 3 and 4.")
+	bob.SendReceipt(t, room3, message3, "m.read")
+	bob.SendReceipt(t, room4, message4, "m.read")
+	time.Sleep(300 * time.Millisecond) // TODO: find a better way to wait until the proxy has processed this.
+
+	syncResp = waitForSyncResponse()
+	t.Log("Bob sees his receipt in room 4, but not in room 3.")
```
Comment (kegsay) [inline tests-e2e/receipts_test.go:325]: So testing for negation I would do what you're doing here and send room 3 BEFORE room 4. That way, you just need to test until you see room 4 and not wait an indeterminate amount of time for room 3, effectively using room 4 as a sentinel.

File: tests-e2e/receipts_test.go [hunk 7]
```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
...
+		select {
+		case res := <-responses:
+			return res
+		case <-time.After(500 * time.Millisecond):
+			t.Fatalf("Timed out waiting for incremental sync response")
+		}
+		return nil
+	}
+
+	go bobIncrementalSync()
```
Comment (kegsay) [inline tests-e2e/receipts_test.go:318]: Same deal as the last PR, you don't need this.

**Review by kegsay — CHANGES_REQUESTED** [review]
A few things about tests, but LGTM otherwise.

File: tests-e2e/receipts_test.go (hunk 2 above)
Comment (DMRobertson) [reply tests-e2e/receipts_test.go:236]: Ugh, forgot a `/g` when I did an `s/1/3`

**Review by kegsay — APPROVED** [review]

**Pull request merged**